### DISCUSSION TITLE
Milvus Empty Segments After Documents deleted

### DISCUSSION THREAD

**Discussion by Harisrini99** [milvus-io/milvus#37603] [Q&A and General discussion]
Hi team,

I’ve deployed Milvus 2.4.11 in a Kubernetes cluster with a large collection of 4 billion documents. I’m currently working on removing unwanted documents using a filter. I deleted 10 million documents through the Milvus filter using the Python Milvus delete client, processing 10 million deletions in batches of 500 with a 10-second pause between each request. The deletions were successful, and search queries are functioning well. However, I noticed that empty segments (segments with 0 rows) in Milvus are not being deleted as expected. Initially, there were 200 empty segments, and after a day, this number reduced to 115. Even after a week, 65 empty segments remain.

`res = self.client.delete(collection_name=collection_name, filter=milvus_filter)`

**Expected behavior:** All empty segments should reach zero.

**Actual behavior:** Empty segments are reducing very slowly, even with high memory and CPU resources allocated.

**Show Segments result for our collection in birdwatcher:**

`show segment --collection 452180062866480917 --state Flushed`

> SegmentID: 453530674841563159 State: Flushed, Row Count:0
SegmentID: 453530674841563166 State: Flushed, Row Count:0
SegmentID: 453530674841563191 State: Flushed, Row Count:0
SegmentID: 453530674841563199 State: Flushed, Row Count:0
SegmentID: 453530674841563206 State: Flushed, Row Count:0
SegmentID: 453530674841563223 State: Flushed, Row Count:0
...
SegmentID: 453530674924430827 State: Flushed, Row Count:0
SegmentID: 452180063744459399 State: Flushed, Row Count:244847
SegmentID: 452180065127316774 State: Flushed, Row Count:258140
SegmentID: 452488476589643128 State: Flushed, Row Count:239264
...

**Comment by yhmo** [comment]
Could you use the latest birdwatcher v1.0.x to show the segment
https://github.com/milvus-io/birdwatcher/releases

The latest birdwatcher can show the segment level.
![Screenshot from 2024-11-12 16-02-41](https://github.com/user-attachments/assets/fd62611c-132a-43e0-aac3-76f2d19b72af)

**Comment by Harisrini99** [comment]
is the new docker image(v1.0.7) pubished ?
i can't see the release in docker registry
https://hub.docker.com/r/milvusdb/birdwatcher/tags

**Reply by yhmo** [reply]
I think the 1.0.7 binary is released: https://github.com/milvus-io/birdwatcher/releases/tag/v1.0.7
No need docker. Just download the tar file(according to your OS) from the "Assets", extract it to disk, and run the binary in the terminal.

**Comment by Harisrini99** [comment]
I checked with the new version and noticed the following output. All the empty segments are at Level L0. Ideally, these segments should be removed over time, but that’s not happening as expected or is occurring very slowly. For instance, after one week, only 130 segments have been removed

> SegmentID: 453530674841563159 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841563166 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841563199 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841563206 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841563223 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841564341 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841568382 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841572290 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841576430 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674841578203 PartitionID: -1 State: Flushed, Level: L0, Row Count:0,  PartitionStatsVersion:0, IsSorted: false 
...
SegmentID: 452180063744459399 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:244847,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452180065127316774 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:258140,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488476589643128 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:239264,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488476650021349 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:243781,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488476768516613 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:255526,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488477111579933 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:255815,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488477134577941 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:202949,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488477174279165 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:210915,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 452488477222029509 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:249257,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674984824274 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:224387,  PartitionStatsVersion:0, IsSorted: false 
SegmentID: 453530674985000473 PartitionID: 452180062866480918 State: Flushed, Level: L1, Row Count:227364,  PartitionStatsVersion:0, IsSorted: false 
...

**Comment by xiaofan-luan** [comment]
@XuanYang-cn 
can you help on answering this question?

**Comment by XuanYang-cn** [comment]
@Harisrini99 Hi, Could you please provide the results of this birdwatcher command, looking for failed and timeout compactions.

```cli
> show compactions --state failed
> show compactions --state timeout
```

### Some general advise:
We're continous enhancing deletion and L0 segment compaction performance now, and we've tested on continuous delete 60% of the 100M entries. In fact, we've made several improvments in v2.4.12, 2.4.13, 2.4.14 and 2.4.15 that could help to ease your problem.

**It's highly recommanded to upgrade Milvus to the latest 2.4.15, upgrading won't be needing to edit the following configs.**

If upgrading is not a choice, there're some other options: 
1. Scale up your DataNde, or scale out DNs for Milvus >=2.4.14, this'll speed up the L0 segment compaction
2. Change config `dataNode.compaction.levelZeroBatchMemoryRatio` from 0.05 to 0.5
3. Add config `dataCoord.compaction.timeout: 7200`

**Comment by Harisrini99** [comment]
Thanks for the reply, we will try upgrading the cluster. Below is the output for show compactions

> Milvus(by-dev) > show compactions --state failed
--- Total compactions:  1652	 Matched compactions:  0
================================================================================
Milvus(by-dev) > 



> Milvus(by-dev) > show compactions --state timeout
--- Total compactions:  1652	 Matched compactions:  0
================================================================================
Milvus(by-dev) >
